lib/c/criu.c

```diff
@@ -1134,29 +1135,65 @@ int criu_set_page_server_address_port(const char *address, int port)
 
 static CriuResp *recv_resp(int socket_fd)
 {
+	struct msghdr msg_hdr = {0};
 	unsigned char *buf = NULL;
-	int len;
+	struct cmsghdr *cmsg;
 	CriuResp *msg = 0;
+	struct iovec io;
+	int len;
 
+	/* Check the size of the waiting data. */
 	len = recv(socket_fd, NULL, 0, MSG_TRUNC | MSG_PEEK);
 	if (len == -1) {
 		perror("Can't read request");
 		goto err;
 	}
 
-	buf = malloc(len);
+	/*
+	 * If there is an FD attached to the protobuf message from CRIU
+	 * the FD will be in the ancillary data. Let's reserve additional
+	 * 32 bytes for that.
+	 */
+	buf = malloc(len+32);
 	if (!buf) {
 		errno = ENOMEM;
 		perror("Can't receive response");
 		goto err;
 	}
 
-	len = recv(socket_fd, buf, len, MSG_TRUNC);
+	io.iov_base = buf;
+	io.iov_len = len;
+	msg_hdr.msg_iov = &io;
+	msg_hdr.msg_iovlen = 1;
+	msg_hdr.msg_control = buf + len;
+	msg_hdr.msg_controllen = 32;
+	len = recvmsg(socket_fd, &msg_hdr, MSG_TRUNC);
+
 	if (len == -1) {
 		perror("Can't read request");
 		goto err;
 	}
 
+	/*
+	 * This will be NULL if no FD is in the message. Currently
+	 * only a response with script set to 'orphan-pts-master'
+	 * has an FD in the ancillary data.
+	 */
+	cmsg = CMSG_FIRSTHDR(&msg_hdr);
+	if (cmsg) {
+		/* We probably got an FD from CRIU. */
+		if (cmsg->cmsg_type != SCM_RIGHTS) {
+			errno = EINVAL;
+			goto err;
+		}
+		/* CTRUNC will be set if msg_hdr.msg_controllen is too small. */
+		if (msg_hdr.msg_flags & MSG_CTRUNC) {
+			errno = ENFILE;
+			goto err;
+		}
+		memcpy(&orphan_pts_master_fd, CMSG_DATA(cmsg), sizeof(int));
```

Comment:
orphan_pts_master_fd = *(int *)CMSG_DATA(cmsg)

Reply:
I initially had it like this and have changed it back to be like this now.

Reply:
Now, thanks to Travis I remember why I did a memcpy. On ARCH=x86_64 I get
```
error: dereferencing type-punned pointer will break strict-aliasing rules
```

Reply:
In this case, we need to have a comment which explains why we use memcpy here.

Reply:
done